### DISCUSSION TITLE
Apply filename attribute to output rather than source block

### DISCUSSION THREAD

**Discussion by royfrancis** [quarto-dev/quarto-cli#7168] [Q&A] [answered]
### Description

````
---
format: html
---

```{r}
#| filename: bash
#| class-output: bash

name <- "demo"
cat(paste("mkdir",name))
```


```{r}
#| filename: bash
#| class-output: bash
#| echo: false

name <- "demo"
cat(paste("mkdir",name))
```
````

![Screenshot 2023-10-07 at 16 56 48](https://github.com/quarto-dev/quarto-cli/assets/5083500/6a5d15db-0851-49ec-b11f-a4746ba0bfc7)

```
$ quarto --version
1.3.433
```

**Comment by mcanouil** [comment, marked as answer]
````qmd
---
format: html
---

```{r}
#| output: asis
#| echo: false
name <- "demo"
cat(
  "```{.bash filename='Bash'}",
  paste("mkdir", name),
  "```",
  sep = "\n"
)
```
````
<img width="1624" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/8896044/fc5e2637-b718-4a52-bf4b-0575bca3e9ed">

**Reply by royfrancis** [reply]
Hacky! :D

**Reply by mcanouil** [reply]
I don't think it's an hack, eventually "advanced usage" of `output: asis` though 🤷‍♂️😎

**Comment by cderv** [comment]
@royfrancis you were not so far with your tries. You just need to use one more knitr option 

`````markdown
```{r}
#| attr-output: "filename='bash'"
#| class-output: bash
#| echo: false
name <- "demo"
cat(paste("mkdir", name))
```
`````

Other provided solution at https://github.com/quarto-dev/quarto-cli/discussions/7168#discussioncomment-7217875 works for sure. 

Here a here a panorama of different way you can achieve this using **knitr** feature

````markdown
---
format: html
keep-md: true
engine: knitr
---

# If you have content you want to show 

```{verbatim}
#| lang: bash
#| attr-source: "filename='bash'"
mkdir demo
```

# If this content is in a file 

```{cat}
#| echo: false 
#| engine-opts: 
#|  file: bash.sh
mkdir demo
```

```{embed}
#| file: bash.sh
#| attr-source: "filename='bash'"
```

```{r}
#| include: false
unlink("bash.sh")
```

# If you want to build content

* Either write in a temp file and use like above 

```{r}
#| include: false
name <- "demo"
tmpfile <- tempfile(fileext = ".sh")
xfun::write_utf8(paste("mkdir",name), tmpfile)
```

```{embed}
#| file: !expr tmpfile
#| attr-source: "filename='bash'"
```

```{r}
#| include: false
unlink(tmpfile)
```

* Or indeed use cat 

```{r}
#| attr-output: "filename='bash'"
#| class-output: bash
#| echo: false
name <- "demo"
cat(paste("mkdir", name))
```

````

![image](https://github.com/quarto-dev/quarto-cli/assets/6791940/d5c28c66-6583-4a4d-b1c8-ad7af0ea7678)

**Reply by royfrancis** [reply]
Thank you! I knew there had to be a more elegant solution.

**Reply by royfrancis** [reply]
Is `attr-output` and similar attributes documented somewhere?

**Reply by mcanouil** [reply]
They are in the reference pages for knitr since "attribute" can only be set via knitr, see <https://quarto.org/docs/reference/cells/cells-knitr.html>.

Note that I opened few months ago and issue to actually allow a more agnostic option comparable to "classes", see:
- https://github.com/quarto-dev/quarto-cli/issues/6253
